google/cloud/spanner_dbapi/cursor.py

```diff
@@ -0,0 +1,333 @@
+        try:
+            return next(self)
+        except StopIteration:
+            return None
+
+    def fetchmany(self, size=None):
+        """Fetch the next set of rows of a query result, returning a sequence
+        of sequences. An empty sequence is returned when no more rows are available.
+
+        :type size: int
+        :param size: (Optional) The maximum number of results to fetch.
+
+        :raises InterfaceError:
+            if the previous call to .execute*() did not produce any result set
+            or if no call was issued yet.
+        """
+        self._raise_if_closed()
+
+        if size is None:
+            size = self.arraysize
+
+        items = []
+        for i in range(size):
+            try:
+                items.append(tuple(self.__next__()))
+            except StopIteration:
+                break
+
+        return items
+
+    def fetchall(self):
+        """Fetch all (remaining) rows of a query result, returning them as
+        a sequence of sequences.
+        """
+        self._raise_if_closed()
+
+        return list(self.__iter__())
+
+    def nextset(self):
+        """A no-op, raising an error if the cursor or connection is closed."""
+        self._raise_if_closed()
+
+    def setinputsizes(self, sizes):
+        """A no-op, raising an error if the cursor or connection is closed."""
+        self._raise_if_closed()
+
+    def setoutputsize(self, size, column=None):
+        """A no-op, raising an error if the cursor or connection is closed."""
+        self._raise_if_closed()
+
+    def _handle_DQL(self, sql, params):
+        with self.connection.database.snapshot() as snapshot:
+            # Reference
+            #  https://googleapis.dev/python/spanner/latest/session-api.html#google.cloud.spanner_v1.session.Session.execute_sql
+            sql, params = parse_utils.sql_pyformat_args_to_spanner(sql, params)
+            res = snapshot.execute_sql(
+                sql, params=params, param_types=get_param_types(params)
+            )
+            if type(res) == int:
+                self._row_count = res
+                self._itr = None
+            else:
+                # Immediately using:
+                #   iter(response)
+                # here, because this Spanner API doesn't provide
+                # easy mechanisms to detect when only a single item
+                # is returned or many, yet mixing results that
+                # are for .fetchone() with those that would result in
+                # many items returns a RuntimeError if .fetchone() is
+                # invoked and vice versa.
+                self._result_set = res
+                # Read the first element so that the StreamedResultSet can
+                # return the metadata after a DQL statement. See issue #155.
+                self._itr = PeekIterator(self._result_set)
+                # Unfortunately, Spanner doesn't seem to send back
+                # information about the number of rows available.
```

Comment:
Yes, unfortunately the row count is sent as part of `ResultSetStats` which is not available until the final `ResultSet` has been received.

See: https://github.com/googleapis/python-spanner/blob/master/google/cloud/spanner_v1/streamed.py#L46